jinahub/encoder/image/clip.py

```diff
@@ -0,0 +1,86 @@
+        if len(current_batch) == batch_size:
+            yield current_batch
+            current_batch = []
+    if current_batch:
+        yield current_batch
+
+
+class CLIPImageEncoder(Executor):
+    """
+    Encode image into embeddings.
+
+    :param model_name: use clip.available_models() to see all available models: ['RN50', 'RN101', 'RN50x4', 'ViT-B/32']
+    :param device: device to use for encoding ['cuda', 'cpu]
+    :param default_batch_size: fallback batch size in case there is not batch size sent in the request
+    :param default_traversal_path: fallback traversal path in case there is not traversal path sent in the request
+    :param jit: Whether to load the optimized JIT model (default) or more hackable non-JIT model.
+    :param channel_axis: axis of the color channel
+    """
+
+    def __init__(
+            self,
+            model_name: str = 'ViT-B/32',
+            device: str = None,
+            default_batch_size: int = 32,
+            default_traversal_path: str = 'r',
+            jit: bool = True,
+            channel_axis: int = 0,
+            *args, **kwargs
+    ):
+        super().__init__(*args, **kwargs)
+        if not device:
+            device = 'cuda' if not device and torch.cuda.is_available() else 'cpu'
+        self.default_batch_size = default_batch_size
+        self.default_traversal_path = default_traversal_path
+        self.channel_axis = channel_axis
+        model, _ = clip.load(model_name, device, jit)
+        self.model = model
+
+    @requests
+    def encode(self, docs: Optional[DocumentArray], parameters: dict, **kwargs):
+        """
+        Encode all docs with images and store the encodings in the embedding attribute of the docs.
+        :param docs: documents sent to the encoder
+        :param parameters: dictionary to define the traversal_
+        :param kwargs:
+        """
+        if docs:
+            document_batches_generator = self._get_input_data(docs, parameters)
+            self._create_embeddings(document_batches_generator)
+
+    def _get_input_data(self, docs: DocumentArray, parameters: dict):
+        traversal_path = parameters.get('traversal_path', self.default_traversal_path)
+        batch_size = parameters.get('batch_size', self.default_batch_size)
+
+        # traverse thought all documents which have to be processed
+        flat_docs = docs.traverse_flat(traversal_path)
+
+        # filter out documents without images
+        filtered_docs = (doc for doc in flat_docs if doc.blob is not None)
+
+        return _batch_generator(filtered_docs, batch_size)
+
+    def _create_embeddings(self, document_batches_generator: Iterable):
+        with torch.no_grad():
+            for document_batch in document_batches_generator:
+                blob_batch = DocumentArray(document_batch).get_attributes('blob')
+                if self.channel_axis != 0:
+                    blob_batch = [np.moveaxis(blob, self.channel_axis, 0) for blob in blob_batch]
```

Comment:
This looks wired for me. The first axis is usually for batching instead of for the color channel.

Reply:
I see why this looks weird to you. 
I traverse through the list and change the change the channel dimension for each ndarray instead of creating one ndarray and swapping the dimension afterwards.
I could change it.

Reply:
I would not swap, but just document